File: capitolo_1/octave/Esempio1_15.m
function Esempio1_15()

  close all
  clear all
  clc
  % Si voglia implementare uno script che secondo l'esempio riporduca i grafici
  % della Figura 1.12, 1.13 (a e b) e 1.14

  %
  % Dati 1.12: Transitori della posizione del processo massa-molla nominale con
  % condizioni iniziali nulle sotto l'azione del controllore. Sono noti i valori
  % nominali Mbar, kbar e hbar (Esempio 1.14)
  %
  Mbar = 1; %Volore scelto di progetto
  hbar = 3; %Volore scelto di progetto
  kbar = 1; %Volore scelto di progetto
  sbar = 1; %Volore scelto di progetto
  dFe = 0; %Ipotesi in cui Fe = Febar
  s0 = 0; %Condizioni iniziali nulle
  s0dot = 0; %Condizioni iniziali nulle
  x0 = [s0 s0dot]; %Vettore condizioni iniziali
  dt = 0.001; %Definizione del tempo
  mus = [0.001 0.5 1 5 10 20]; % Valori di mu per fare il grafico

  % Figura 1.12
  t = 0:dt:10; %Vettore tempo
  risultatiControllore = zeros([length(t), length(mus)]);
  legends = {};
  for i = 1 : length(mus)
    [t,x] = ode45(@controlloreAnelloChiuso, t, [x0], Mbar, hbar, kbar, dFe, sbar, mus(i));
    risultatiControllore(:,i) = x(:,1);
    legends{i} = ['\mu = ' num2str(mus(i))];
  endfor

  figure(1)
  plot(t,risultatiControllore, [0 10], [1 1], 'k');
  grid;
  xlabel('Tempo (t)')
  ylabel('Stato s')
  title('Evoluzione dello stato s per diversi guadagni (\mu) e C.I. nulle')
  legend(legends,'Location','SouthEast');

  %
  % Dati 1.13a: Transitori della posizione del processo massa-molla nominale con
  % condizioni iniziali non nulle per s0 e s0dot sotto l'azione del controllore.
  %
  Mbar = 1; %Volore scelto di progetto
  hbar = 3; %Volore scelto di progetto
  kbar = 1; %Volore scelto di progetto
  sbar = 0; %Volore scelto di progetto
  dFe = 0; %Ipotesi in cui Fe = Febar
  dt = 0.001; %Definizione del tempo
  s0 = 1; %Condizioni iniziali non nulle
  s0dot = 1; %Condizioni iniziali non nulle
  x0 = [s0 s0dot]; %Vettore condizioni iniziali

  % Figura 1.13a
  t = 0:dt:10; %Vettore tempo
  mus = [0.001 0.5 1 5 10 20]; % Valori di mu per fare il grafico
  risultatiControllore = zeros([length(t), length(mus)]);
  legends = {};
  for i = 1 : length(mus)
    [t,x] = ode45(@controlloreAnelloChiuso, t, [x0], Mbar, hbar, kbar, dFe, sbar, mus(i));
    risultatiControllore(:,i) = x(:,1);
    legends{i} = ['\mu = ' num2str(mus(i))];
  endfor

  figure(2)
  plot(t,risultatiControllore, [0 10], [0 0], 'k');
  grid;
  xlabel('Tempo (t)')
  ylabel('Stato s')
  title('Evoluzione dello stato s per diversi guadagni (\mu) e C.I. non nulle')
  legend(legends,'Location','SouthEast');

  %
  % Dati 1.13b: Transitori della posizione del processo massa-molla nominale con
  % condizioni iniziali nulle per s0 e s0dot, Fe!=Febar sotto l'azione del controllore.
  %
  Mbar = 1; %Volore scelto di progetto
  hbar = 3; %Volore scelto di progetto
  kbar = 1; %Volore scelto di progetto
  sbar = 0; %Volore scelto di progetto
  dFe = 1; %Ipotesi in cui Fe = Febar
  dt = 0.001; %Definizione del tempo
  s0 = 0; %Condizioni iniziali non nulle
  s0dot = 0; %Condizioni iniziali non nulle
  x0 = [s0 s0dot]; %Vettore condizioni iniziali

  % Figura 1.13b
  t = 0:dt:10; %Vettore tempo
  mus = [0.001 0.5 1 5 10 20]; % Valori di mu per fare il grafico
  risultatiControllore = zeros([length(t), length(mus)]);
  legends = {};
  for i = 1 : length(mus)
    [t,x] = ode45(@controlloreAnelloChiuso, t, [x0], Mbar, hbar, kbar, dFe, sbar, mus(i));
    risultatiControllore(:,i) = x(:,1);
    legends{i} = ['\mu = ' num2str(mus(i))];
  endfor

  figure(3)
  plot(t,risultatiControllore, [0 10], [0 0], 'k');
  grid;
  xlabel('Tempo (t)')
  ylabel('Stato s')
  title('Evoluzione dello stato s per diversi guadagni (\mu) e C.I. non nulle')
  legend(legends,'Location','SouthEast');

  %
  % Dati 1.14: Transitori della forza motrice del processo massa-molla nominale con
  % condizioni iniziali nulle per s0 e s0dot sotto l'azione del controllore.
  % Dall'esempio 1.14 risulta che
  %   Fm = kbar * sbar + Febar + mu * e
  % dove e = sbar - s
  %
  Mbar = 1; %Volore scelto di progetto
  hbar = 3; %Volore scelto di progetto
  kbar = 1; %Volore scelto di progetto
  sbar = 1; %Volore scelto di progetto
  dFe = 0; %Ipotesi in cui Fe = Febar
  dt = 0.001; %Definizione del tempo
  s0 = 0; %Condizioni iniziali non nulle
  s0dot = 0; %Condizioni iniziali non nulle
  x0 = [s0 s0dot]; %Vettore condizioni iniziali

  % Figura 1.14
  t = 0:dt:10; %Vettore tempo
  mus = [0.001 0.5 1 5 10 20]; % Valori di mu per fare il grafico
  transitoriFm = zeros([length(t), length(mus)]);
  legends = {};
  for i = 1 : length(mus)
    [t,x] = ode45(@controlloreAnelloChiuso, t, [x0], Mbar, hbar, kbar, dFe, sbar, mus(i));
    s = x(:,1);
    transitoriFm(:,i) = kbar * sbar + dFe + mus(i) * (sbar - s);
    legends{i} = ['\mu = ' num2str(mus(i))];
  endfor

  figure(4)
  plot(t,transitoriFm, [0 10], [0 0], 'k');
  grid;
  xlabel('Tempo (t)')
  ylabel('Forza motrice')
  title('Transitorio Fm per diversi guadagni (\mu) e C.I. nulle')
  legend(legends,'Location','SouthEast');

endfunction

function [xp] = controlloreAnelloChiuso(t, x, M, h, k, dFe, s, mu)
  % Questa funzione implementa l'evoluzione di stato per l'esempio descritto.
  % Data l'equazione
  %   M * sddot(t) = -h * sdot(t) - (k+mu)*s(t) + (kbar + mu)*sbar - Fe + Febar
  % Si considera Febar - Fe = dFe (delta Fe). Inoltre si assume lo stato come
  %   x(t) = [ s(t)  sdot(t) ]
  % Pertanto l'evoluzione di stato sarà
  %   xdot(t) = [ sdot(t) ((-h/M)*sdot(t) - ((k+mu)/M)*s(t) + ((kbar + mu)/M)*sbar + dFe/M)]
  xp = [x(2) ...
        -(h/M) * x(2) - ((k + mu)/M * x(1)) + ((k + mu)/M) * s + dFe/M];
endfunction

File: capitolo_1/cpp/Esempio1_15.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <string>
#include <map>
#include "matplotlibcpp.h"

namespace plt = matplotlibcpp;

// --------------------------
// Parametri di sistema
// --------------------------
struct Params {
    double Mbar;
    double hbar;
    double kbar;
    double dFe;
    double sbar;
    double mu;
};

// --------------------------
// Equazione differenziale - Controllore ad anello chiuso
// --------------------------
std::vector<double> f_anelloChiuso(double t, const std::vector<double>& x, const Params& p) {
    (void)t; // unused
    std::vector<double> dx(2);
    dx[0] = x[1];
    dx[1] = -(p.hbar/p.Mbar) * x[1] - ((p.kbar + p.mu)/p.Mbar * x[0]) + ((p.kbar + p.mu)/p.Mbar) * p.sbar + p.dFe/p.Mbar;
    return dx;
}

// --------------------------
// Operatori utili
// --------------------------
std::vector<double> operator+(const std::vector<double>& a, const std::vector<double>& b) {
    std::vector<double> r(a.size());
    for (size_t i = 0; i < a.size(); ++i) r[i] = a[i] + b[i];
    return r;
}
std::vector<double> operator*(const std::vector<double>& a, double s) {
    std::vector<double> r(a.size());
    for (size_t i = 0; i < a.size(); ++i) r[i] = a[i] * s;
    return r;
}

// --------------------------
// Runge–Kutta 4° ordine
// --------------------------
template <typename Func>
std::vector<std::vector<double>> rungeKutta4(Func f, std::vector<double> x0,
                                             double t0, double tf, double dt, const Params& p) {
    int N = static_cast<int>((tf - t0) / dt);
    std::vector<std::vector<double>> result;
    result.reserve(N + 1);

    std::vector<double> x = x0;
    double t = t0;
    result.push_back(x);

    for (int i = 0; i < N; ++i) {
        auto k1 = f(t, x, p);
        auto k2 = f(t + dt / 2, x + k1 * (dt / 2), p);
        auto k3 = f(t + dt / 2, x + k2 * (dt / 2), p);
        auto k4 = f(t + dt, x + k3 * dt, p);

        x = x + (k1 + k2 * 2 + k3 * 2 + k4) * (dt / 6);
        t += dt;
        result.push_back(x);
    }

    return result;
}

// --------------------------
// MAIN
// --------------------------
int main() {
    // Dati
    double Mbar = 1, hbar = 3, kbar = 1, sbar = 1, dFe = 0;
    std::vector<double> mus_prop = {0.001, 0.5, 1.0, 5.0, 10.0, 20.0};
    double dt = 0.001;
    double t0 = 0, tf = 10.0;
    int N = static_cast<int>((tf - t0) / dt);
    std::vector<double> time(N + 1);
    for (int i = 0; i <= N; ++i) time[i] = t0 + i * dt;

    // -----------------------------------------
    // Figura 1.12 - Transitorio posizione massa-molla, C.I. nulle, sotto azione del controllore.
    // -----------------------------------------
    std::map<double, std::vector<double>> risultatiControllore;
    for (auto mu : mus_prop) {
        Params p{Mbar, hbar, kbar, dFe, sbar, mu};
        auto traj = rungeKutta4(f_anelloChiuso, {0.0, 0.0}, t0, tf, dt, p);
        std::vector<double> C;
        for (auto& x : traj) C.push_back(x[0]);
        risultatiControllore[mu] = C;
    }

    // Plot Figura 1.12
    plt::figure(1);
    for (auto& [mu, C] : risultatiControllore)
        plt::plot(time, C, {{"label", "μ = " + std::to_string(mu)}});
    plt::xlabel("Tempo (t)");
    plt::ylabel("Posizione s");
    plt::title("Transitorio della posizione - C.I. nulle");
    plt::legend();
    plt::grid(true);

    // -----------------------------------------
    // Figura 1.13 - Transitorio posizione massa-molla sotto azione del controllore con sbar = 0.
    // a) C.I. non nulle
    // b) C.I. nulle e dFe != 0
    // -----------------------------------------
    risultatiControllore.clear();
    sbar = 0.0;
    double s0 = 1.0;
    double v0 = 1.0;
    for (auto mu : mus_prop) {
        Params p{Mbar, hbar, kbar, dFe, sbar, mu};
        auto traj = rungeKutta4(f_anelloChiuso, {s0, v0}, t0, tf, dt, p);
        std::vector<double> C;
        for (auto& x : traj) C.push_back(x[0]);
        risultatiControllore[mu] = C;
    }

    // Plot Figura 1.13a
    plt::figure(2);
    for (auto& [mu, C] : risultatiControllore)
        plt::plot(time, C, {{"label", "μ = " + std::to_string(mu)}});
    plt::xlabel("Tempo (t)");
    plt::ylabel("Posizione s");
    plt::title("Transitorio della posizione - C.I. non nulle");
    plt::legend();
    plt::grid(true);

    risultatiControllore.clear();
    dFe = 1.0;
    for (auto mu : mus_prop) {
        Params p{Mbar, hbar, kbar, dFe, sbar, mu};
        auto traj = rungeKutta4(f_anelloChiuso, {0.0, 0.0}, t0, tf, dt, p);
        std::vector<double> C;
        for (auto& x : traj) C.push_back(x[0]);
        risultatiControllore[mu] = C;
    }

    // Plot Figura 1.13b
    plt::figure(3);
    for (auto& [mu, C] : risultatiControllore)
        plt::plot(time, C, {{"label", "μ = " + std::to_string(mu)}});
    plt::xlabel("Tempo (t)");
    plt::ylabel("Posizione s");
    plt::title("Transitorio della posizione - C.I. nulle, Fe ≠ Febar");
    plt::legend();
    plt::grid(true);

    // -----------------------------------------
    // Figura 1.14 - Transitorio forza motricesotto azione del controllore con C.I. nulle.
    // -----------------------------------------
    std::map<double, std::vector<double>> risultatiFmotrice;
    sbar = 1.0;
    dFe = 0.0;
    for (auto mu : mus_prop) {
        Params p{Mbar, hbar, kbar, dFe, sbar, mu};
        auto traj = rungeKutta4(f_anelloChiuso, {0.0, 0.0}, t0, tf, dt, p);
        std::vector<double> C;
        for (auto& x : traj){
            double Fm = kbar * sbar + dFe + mu * (sbar - x[0]);
            C.push_back(Fm);
        }
        risultatiFmotrice[mu] = C;
    }

    // Plot Figura 1.14
    plt::figure(4);
    for (auto& [mu, C] : risultatiFmotrice)
        plt::plot(time, C, {{"label", "μ = " + std::to_string(mu)}});
    plt::xlabel("Tempo (t)");
    plt::ylabel("Posizione s");
    plt::title("Transitorio della Fm - C.I. nulle");
    plt::legend();
    plt::grid(true);

    plt::show();

    return 0;
}
